### ISSUE TITLE
Allow aws cloudformation deploy to use template-url

### ISSUE THREAD

**Issue by armaseg** [aws/aws-cli#2461]
As the [How to create serverless applications using AWS SAM](https://github.com/awslabs/serverless-application-model/blob/master/HOWTO.md) says : 

> SAM template is deployed to AWS CloudFormation by creating a changeset using the SAM template followed by executing the changeset... ...Alternatively, you can use aws cloudformation deploy CLI command to deploy the SAM template. Under-the-hood it creates and executes a changeset and waits until the deployment completes.

Nevertheless, a Cloudformation [create-change-set](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/create-change-set.html) allows to use the `--template-url` parameter, however the `aws cloudformation deploy` command does not. Please update the CLI to support `--template-url` while doing a `deploy`.

This is a feature already available on the AWS Management Console, where  you can deploy your stack using the URL that points to a template that is located in an S3 bucket.

AWS CLI version: 1.11.46

**Comment by kyleknap** [comment]
Marking as a feature request. @sanathkr Any thoughts on this?

**Comment by sanathkr** [comment]
Possible, but the problem is `deploy` command does a bit more than create-change-set. It allows you to specify only certain parameters that you'd like to override, instead of all the parameters in your template. It reads your template and passes UsePreviousValue=True for parameters that have not been overridden. If your template is in S3, we have to pull it down, which the CLI might not have permissions for.

**Comment by jghaines** [comment]
The lack of `--template-url` support means that `aws cloudformation deploy` templates are limited to 51200 bytes in size. 

Even if `--template-url` wouldn't allow UsePreviousValue that would be a useful feature.

**Comment by matbos** [comment]
Is there any timeline for this feature or a workaround available(other than splitting the template)? My template is affected by the limit.

**Comment by jghaines** [comment]
One workaround is to split the template.

Alternatively, you can rewrite the deploy logic using AWS CLI.  You just need to:

- Check for the existence of the stack `aws cloudformation describe-stacks` - and parse the stderr for "does not exist"
- Create a change-set `aws cloudformation create-change-set` - using --change-set-type CREATE or UPDATE appropriately
- Wait for creation to complete `aws cloudformation wait change-set-create-complete`
- Do an `aws cloudformation describe-change-set` and check the StatusReason for "The submitted information didn't contain changes"
- Conditionally execute the change-set `aws cloudformation execute-change-set`. Don't execute no-op changesets.
- Wait for the change-set to execute. `aws cloudformation wait change-set-create-complete` waits for the CREATE_COMPLETE  status and behaves badly in case of failure (takes 1 hour to time-out). Instead, write a loop to poll `aws cloudformation describe-stacks` for the stack status. There are [16 possible statuses](https://github.com/rvedotrc/aws-cloudformation-stack-states)

**Comment by lox** [comment]
The lack of this also makes it hard to follow https://aws.amazon.com/blogs/devops/aws-cloudformation-security-best-practices/ and restrict access to a specific template.

**Comment by sanathkr** [comment]
Yeah, this is definitely an issue we should fix. this requires quite a bit of changes to the code. PRs appreciated!

Instead of reading a local file, AWSCLI will pull the template from given S3 location, parse the parameters out, merge with the parameter overrides arguments, and call create-change-set with S3 template URL instead of uploading the template text

**Comment by iDVB** [comment]
@stealthycoin Where did things leave off with this? It looks like it was just closed, but I still don't see support for `--template-url` in the latest cli version. What am I missing?

**Comment by galartsi-cg** [comment]
Can someone reopen this? it is not fixed.

**Comment by echernyavskiy** [comment]
Could this be re-opened please? Lack of this option prevents us from using nested stacks for larger applications for example.

**Comment by tobilg** [comment]
I think this works via the `--s3-bucket`, `--s3-prefix` if the template is uploaded to a S3 bucket beforehand (see https://docs.aws.amazon.com/cli/latest/reference/cloudformation/deploy/index.html)

**Comment by timharris777** [comment]
This is not fixed and needs to be re-opened. `aws cloudformation deploy` should be able to run against a template in s3 without having a local file.

**Comment by timharris777** [comment]
I just submitted a PR with working code that accepts either a local path or s3 url to the cloudformation template with the `--template-file` option.

https://github.com/aws/aws-cli/pull/6409

**Comment by HristoStoyanovMM** [comment]
I am surprised to see that, nobody mentioned the following aws cloudformation deploy arguments:
```
--s3-bucket (string) The name of the S3 bucket where this command uploads your CloudFormation template. This is required the deployments of templates sized greater than 51,200 bytes

--s3-prefix (string) A prefix name that the command adds to the artifacts’ name when it uploads them to the S3 bucket. The prefix name is a path name (folder name) for the S3 bucket.
```
Maybe those were introduced after this issue had been opened. I was about to miss them as I was searching for --template-url argument because of how aws cloudformate create-change-set works, and because the PR above is still open.
